### PR TITLE
Onboarding for Market Process Analyzer to AxonIvy Market

### PR DESCRIPTION
**Pull request by ntqdinh-axonivy**

### CODE REVIEW

**Comment by ntqdinh-axonivy** [comment]
Hi @trungmaihova @ntnchuong 
I have folked your dev branch to this PR before we can public market process analyzer to marketplace.
I also raise some feedback before we can release it, please check and do not hesitate to discuss with us if you feel some feedback can be ignore.
Btw, could you please confirm that the product name would be "Market Process Analyzer" or "Process Analyzer" ?
Thank you so much.

File: process-analyzer-demo/processes/Analysis Processes/FlowDemoBasic.p.json [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+      },
+      "connect" : [
+        { "id" : "f13", "to" : "f12", "label" : {
+            "name" : [
+              "{internal}",
+              "{external}",
+              "{mixed}"
+            ]
+          } },
+        { "id" : "f15", "to" : "f7", "color" : "default path" }
```
[the remark this answers is not in the document]
Comment (ntnchuong) [reply process-analyzer-demo/processes/Analysis Processes/FlowDemoBasic.p.json:76]: actually we did it on purpose for test case exception when having two empty condition

File: process-analyzer-test/src/com/axonivy/utils/process/analyzer/test/UseCase.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.axonivy.utils.process.analyzer.test;
+
+
+public enum UseCase {
```
[the remark this answers is not in the document]
Comment (ntnchuong) [reply process-analyzer-test/src/com/axonivy/utils/process/analyzer/test/UseCase.java:4]: library provide flexible enum type => let user define it by their own

File: process-analyzer/src/com/axonivy/utils/process/analyzer/helper/DateTimeHelper.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import ch.ivyteam.ivy.application.calendar.IBusinessCalendar;
+import ch.ivyteam.ivy.environment.Ivy;
+import ch.ivyteam.ivy.scripting.objects.BusinessDuration;
+
+public class DateTimeHelper {
+	private static final IBusinessCalendar GERMANY_CALENDAR = Ivy.cal().get("Germany");
+
+	public static Date getBusinessStartTimestamp(Date startJavaDateTime) {
+		ch.ivyteam.ivy.scripting.objects.DateTime startDateTime = new ch.ivyteam.ivy.scripting.objects.DateTime(startJavaDateTime);
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer/src/com/axonivy/utils/process/analyzer/helper/DateTimeHelper.java:14]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+		result.setInternalPaths(paths);
+		
+		return result;
+	}
+	
+	/**
+	 * Find path on sub process
+	 */	
+	private List<AnalysisPath> findPathOfSubProcess(ProcessElement subProcessElement, String flowName, FindType findType) throws Exception {
+		// find start element EmbeddedProcessElement subProcessElement.getEmbeddedProcess()
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java:477]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+		
+		if(isEmpty(sequenceFlow.getEdge().getLabel().getText())) {
+			return true;
+		};
+
+		return false;
+	}
+
+	private boolean hasFlowName(SequenceFlow sequenceFlow, String flowName) {
+		String label = sequenceFlow.getEdge().getLabel().getText();
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java:576]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer/src/com/axonivy/utils/process/analyzer/internal/WorkflowPath.java [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+		String elementName = outcome.getName();
+		
+		DetectedElement element = new DetectedElement(pid, elementName);
+		return element;
+	}
+	
+	private DetectedElement createDetectedTask(SingleTaskCreator task, Enum<?> useCase, Duration timeUntilStartAt) {
+		return createDetectedTask(task, task.getTaskConfig(), useCase, timeUntilStartAt);
+	}
+	private DetectedElement createDetectedTask(TaskAndCaseModifier task, TaskConfig taskConfig, Enum<?> useCase, Duration timeUntilStartAt) {
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer/src/com/axonivy/utils/process/analyzer/internal/WorkflowPath.java:302]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer-test/src/com/axonivy/utils/process/analyzer/test/UseCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer-test/src/com/axonivy/utils/process/analyzer/test/UseCase.java:4]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer-test/pom.xml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+		<plugins>
+			<plugin>
+				<groupId>com.axonivy.ivy.ci</groupId>
+				<artifactId>project-build-plugin</artifactId>
+				<version>${build.plugin.version}</version>
+				<extensions>true</extensions>
+			</plugin>
+		</plugins>
+	</build>
+</project>
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer-test/pom.xml:35]: https://1ivy.atlassian.net/browse/TE-579

File: process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+		}
+		return intersectNodes.entrySet().stream()
+				.sorted(Map.Entry.<ProcessElement, Set<ProcessElement>>comparingByValue((Set<ProcessElement> a, Set<ProcessElement> b) -> Integer.compare(a.size(), b.size())).reversed())
+				.collect(toMap(Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> oldValue, LinkedHashMap::new));
+	}
+
+	/**
+	 * Using Recursion Algorithm To Find Tasks On Graph.
+	 */
+	private List<AnalysisPath> findPath(ProcessElement from, String flowName, FindType findType, List<AnalysisPath> currentPath) throws Exception {
```
[the remark this answers is not in the document]
Comment (trungmaihova) [reply process-analyzer/src/com/axonivy/utils/process/analyzer/internal/PathFinder.java:316]: it is main function with recursion calling. We can move block code from L358->L362 to a function but I think we should keep to easy understand.

**Comment by ivy-tru** [comment]
Hi @ntqdinh-axonivy 
Thanks for your review comments.

>  Btw, could you please confirm that the product name would be "Market Process Analyzer" or "Process Analyzer" ?

Official name of the market artifact should be "Advanced Process Analyzer". This name should already be used in the documentation files. It changed two times during the project, so sometimes we accidentally use the wrong name.

**Comment by ivy-rew** [comment]
are we still waiting for anything @ivy-tru @ntnchuong @trungmaihova ? 
I'd be happy to have a first version integrated asap.
As we have a build-pipeline that validates market-repos; and fails due to the 'virigin' state within the initial repo commit.

**Comment by ntqdinh-axonivy** [comment]
> are we still waiting for anything @ivy-tru @ntnchuong @trungmaihova ? 
> I'd be happy to have a first version integrated asap.
> As we have a build-pipeline that validates market-repos; and fails due to the 'virigin' state within the initial repo commit.
> There's no need to enforce a publishing; just to merge this state to 'master' would help us already.
> 

Hi @ivy-rew 
There is still a small issue of the group id in the test module's POM file left before merging. We are in progress and will fix it soon tomorrow.
I'm sorry for our delay.

**Comment by ntqdinh-axonivy** [comment]
Hi @trungmaihova @ntnchuong 
FYI @ivy-rew @ivy-rew @ivy-sgi 
I would like to close this PR because the origin repo closed the folk connection and did not update the latest commit.
The follow-up PR is created https://github.com/axonivy-market/advanced-process-analyzer/pull/2 for onboarding the advanced process analyzer utils to market.

**Pull request closed without merging**